src/coreclr/jit/importer.cpp

```diff
@@ -7210,7 +7208,19 @@ void Compiler::impImportBlockCode(BasicBlock* block)
                 // The array helper takes a native int for array length.
                 // So if we have an int, explicitly extend it to be a native int.
                 index = impImplicitIorI4Cast(index, TYP_I_IMPL);
-                op1   = gtNewHelperCallNode(CORINFO_HELP_LDELEMA_REF, TYP_BYREF, arr, index, type);
+
+                if ((gtGetArrayElementClassHandle(arr) == ldelemClsHnd) && impIsClassExact(ldelemClsHnd))
```

Comment:
I'm assuming that unverifiable/unsafe code, or specifically code with runtime behavior that actually deviates from the static types, has undefined behavior.  In other words, we don't need to throw the exception on an object[] that has been misrepresented as a string[] and then a ldelem.a string performed on it.  (This makes sense since such optimizations would then be impossible; it's just against a strict reading of the specification.)

Reply:
> Shouldn't this also be conditional on `opts.OptimizationEnabled()` ? I see you had a minopts diff...

Addressed

Reply:
> In other words, we don't need to throw the exception on an object[] that has been misrepresented as a string[] and then a ldelem.a string performed on it.

I'm not sure what is going to happen (and what is expected) in this case so I'm only optimizing a valid case

Reply:
I assume this case should fail on the `stloc` step

Reply:
Oh, do we reject (at least some kinds of) unverifiable code now?  This is just moving an object reference around; it's only if one looks at the so-called verification types does it break down.  (I guess I should just try it.. I'd be happy to learn this.)

This definitely could be done with Unsafe.As, but the docs there are clear about improper use being undefined.